ganttproject/src/main/java/net/sourceforge/ganttproject/storage/SqlProjectDatabaseImpl.kt

```diff
@@ -127,34 +148,51 @@ class SqlProjectDatabaseImpl(private val dataSource: DataSource) : ProjectDataba
   }
 
   @Throws(ProjectDatabaseException::class)
-  override fun fetchLogRecords(startId: Int, limit: Int): List<LogRecord> = withDSL({ "Failed to fetch log records" }) { dsl ->
+  override fun startTransaction() {
+    if (currentTxn != null) throw ProjectDatabaseException("Previous transaction not committed")
+    currentTxn = mutableListOf()
+  }
+
+  @Throws(ProjectDatabaseException::class)
+  override fun commitTransaction() {
+    try {
+      executeAndLog(currentTxn ?: throw ProjectDatabaseException("No transaction started"))
+    } finally {
+      currentTxn = null
+    }
+  }
+
+  @Throws(ProjectDatabaseException::class)
+  override fun fetchTransactions(startTxnId: Int, limit: Int): List<XlogRecord> = withDSL({ "Failed to fetch log records" }) { dsl ->
     dsl
       .selectFrom(LOGRECORD)
-      .where(LOGRECORD.ID.ge(startId))
-      .orderBy(LOGRECORD.ID.asc())
-      .limit(limit)
-      .map { LogRecord(it.id, it.sqlStatement) }
+      .where(LOGRECORD.TXN_ID.ge(startTxnId).and(LOGRECORD.TXN_ID.lt(startTxnId + limit)))
+      .orderBy(LOGRECORD.TXN_ID, LOGRECORD.ID)
+      .fetchGroups(LOGRECORD.TXN_ID, LOGRECORD.SQL_STATEMENT)
+      .toList()
+      .sortedBy { it.first }
+      .map { XlogRecord(it.second) }
```

Comment:
I don't think that you need to apply this toList and subsequent sorting. It is not going to be expensive, but makes the idea more difficult to understand. If you want to map the collection of list of statements, which is produced by fetchGroups to a list of XlogRecord, you can use `.values.map { XlogRecord(it) }`

Reply:
That's nicer, thanks!